AX: page sometimes points to wrong isolated tree with ENABLE_ACCESSIBILITY_LOCAL_FRAME=1

minorninth opened the pull request:
#### cba49fa3cb9892ba98026665997eae10dd3213d1
<pre>
AX: page sometimes points to wrong isolated tree with ENABLE_ACCESSIBILITY_LOCAL_FRAME=1
<a href="https://bugs.webkit.org/show_bug.cgi?id=299705">https://bugs.webkit.org/show_bug.cgi?id=299705</a>
<a href="https://rdar.apple.com/161522668">rdar://161522668</a>

Reviewed by Tyler Wilcock.

An isolated tree is registering itself with
WKAccessibilityWebPageObjectBase, but if there&apos;s more than one
isolated tree (when ENABLE_ACCESSIBILITY_LOCAL_FRAME=1, which will be
needed for site isolation) then only the main frame should be
registered. Without this fix, VoiceOver sometimes lands on an iframe with no
way to get to the main page.

Not covered by tests because they seem to go through _WKAccessibilityRootObjectForTesting
instead.

* Source/WebKit/WebProcess/WebPage/mac/WKAccessibilityWebPageObjectBase.mm:
(-[WKAccessibilityWebPageObjectBase setIsolatedTree:]):

Canonical link: <a href="https://commits.webkit.org/301150@main">https://commits.webkit.org/301150@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d07da1845f44a4bd93a500e7588cfda7aa942a69

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/125099 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/44766 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/35505 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/131947 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/76970 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/45457 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/53330 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/95230 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/76970 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/128053 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/36290 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/111872 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/75772 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/35196 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/75428 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/106055 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/30261 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/134626 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/51911 "Built successfully") | [⏳ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/macOS-Sequoia-Debug-WK2-Tests-EWS "Waiting to run tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/103698 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/52343 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/108089 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/103470 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26345 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/48813 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/27104 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/48963 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/51802 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/57589 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/51170 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/54526 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/52860 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on Source/WebKit/WebProcess/WebPage/mac/WKAccessibilityWebPageObjectBase.mm:
```diff
@@ -197,6 +197,26 @@ - (void)setIsolatedTree:(Ref<WebCore::AXIsolatedTree>&&)tree
         return;
     }
+
+    auto *cache = tree->axObjectCache();
```
The safer-cpp bot may tell us to use a `CheckedPtr` for this AXObjectCache instead of a raw pointer. And I think in any case, typical WebKit code style for C++ has the `*` connected to `auto` — `auto* cache`

twilco commented on Source/WebKit/WebProcess/WebPage/mac/WKAccessibilityWebPageObjectBase.mm:
```diff
@@ -197,6 +197,26 @@ - (void)setIsolatedTree:(Ref<WebCore::AXIsolatedTree>&&)tree
+        return;
+
+    auto optionalFrameIDFromIsolatedTree = cache->frameID();
+    if (!optionalFrameIDFromIsolatedTree)
```
Is this local variable name accurate? Seems like we're getting this frame ID from the cache? Also, maybe rather than having `optional` in the name, what if we just specified the type instead of using auto?

`std::optional frameID = cache->frameID();`

minorninth replied:
Specifying the type is cleaner, thanks!

twilco commented on Source/WebKit/WebProcess/WebPage/mac/WKAccessibilityWebPageObjectBase.mm:
```diff
@@ -197,6 +197,26 @@ - (void)setIsolatedTree:(Ref<WebCore::AXIsolatedTree>&&)tree
+
+    auto* mainFrame = m_page->mainFrame();
+    if (!mainFrame)
+        return;
```
A bit shorter way to write this:

```cpp
auto* mainFrame = m_page ? m_page->mainFrame() : nullptr;
if (!mainFrame)
    return;
```

twilco commented on Source/WebKit/WebProcess/WebPage/mac/WKAccessibilityWebPageObjectBase.mm:
```diff
@@ -197,6 +197,26 @@ - (void)setIsolatedTree:(Ref<WebCore::AXIsolatedTree>&&)tree
+    if (!mainFrame)
+        return;
+
+    // Ignore an isolated tree that's not the main frame.
```
It's nice to mention this. It may also be helpful to explain what bad thing would happen if we didn't ignore non-main-frame isolated trees

minorninth committed:
AX: page sometimes points to wrong isolated tree with ENABLE_ACCESSIBILITY_LOCAL_FRAME=1
https://bugs.webkit.org/show_bug.cgi?id=299705
rdar://161522668

Reviewed by Tyler Wilcock.

An isolated tree is registering itself with
WKAccessibilityWebPageObjectBase, but if there's more than one
isolated tree (when ENABLE_ACCESSIBILITY_LOCAL_FRAME=1, which will be
needed for site isolation) then only the main frame should be
registered. Without this fix, VoiceOver sometimes lands on an iframe with no
way to get to the main page.

Not covered by tests because they seem to go through _WKAccessibilityRootObjectForTesting
instead.

* Source/WebKit/WebProcess/WebPage/mac/WKAccessibilityWebPageObjectBase.mm:
(-[WKAccessibilityWebPageObjectBase setIsolatedTree:]):

Canonical link: https://commits.webkit.org/301150@main
